components/server/ee/src/gitlab/gitlab-app-support.ts

```diff
@@ -0,0 +1,51 @@
+/**
+ * Copyright (c) 2020 Gitpod GmbH. All rights reserved.
+ * Licensed under the Gitpod Enterprise Source Code License,
+ * See License.enterprise.txt in the project root folder.
+ */
+
+import { ProviderRepository, User } from "@gitpod/gitpod-protocol";
+import { inject, injectable } from "inversify";
+import { TokenProvider } from "../../../src/user/token-provider";
+import { UserDB } from "@gitpod/gitpod-db/lib";
+import { Gitlab } from "@gitbeaker/node";
+
+@injectable()
+export class GitLabAppSupport {
+
+    @inject(UserDB) protected readonly userDB: UserDB;
+    @inject(TokenProvider) protected readonly tokenProvider: TokenProvider;
+
+    async getProviderRepositoriesForUser(params: { user: User, provider: string, hints?: object }): Promise<ProviderRepository[]> {
```

Comment:
**issue**: This is probably minor for this iteration, as this is an edge case, but it takes 5-6 seconds for me to load the next step. FWIW, I'm a member of ~30 groups. In the future a loading indicator could help.

Reply:
Thank for the hint. Just addmin `perPage: 100` to override the default value of `20`. That should be significant improvement in your case.

Reply:
Thanks @AlexTugarev! Seems 2x faster! FWIW, Still taking 2-3 seconds.

Reply:
Awesome. How many projects do you see there? According to the logs, each API response takes ~600ms, which is really slow.

Reply:
Most groups contain less than 10 projects. For the personal account _all_ projects (38) are listed.

**thought**: I think the delay happens while fetching all groups or the ones with proper permissions and maybe because the account is a direct member of the `gitlab-org` group which contains 300+ second-level groups. Tried this with a test account and groups and projects loaded in under 1 second. 🎉